lib/block_view_message.go

```diff
@@ -79,86 +178,172 @@ func (bav *UtxoView) deleteMessageMappings(message *PGMessage) {
-	// This fills in any gaps in the view so that, after this, the view should contain
-	// the union of what it had before plus what was in the db.
-	for _, dbMessageEntry := range dbMessageEntries {
-		messageKey := MakeMessageKey(publicKey, dbMessageEntry.TstampNanos)
-		bav._getMessageEntryForMessageKey(&messageKey)
-	}
+	// This is our helper map to keep track of all user messaging keys.
+	messagingKeysMap := make(map[PublicKey]*MessagingKeyEntry)
 
-	// Now that the view mappings are a complete picture, iterate through them
-	// and set them on the map we're returning. Skip entries that don't match
-	// our public key or that are deleted. Note that only considering mappings
-	// where our public key is part of the key should ensure there are no
-	// duplicates in the resulting list.
-	messageEntriesToReturn := []*MessageEntry{}
-	for viewMessageKey, viewMessageEntry := range bav.MessageKeyToMessageEntry {
-		if viewMessageEntry.isDeleted {
+	// Start by fetching all the messages we have in the UtxoView.
+	for _, messagingKeyEntry := range bav.MessagingKeyToMessagingKeyEntry {
+		// We don't check for deleted entries now, we will do that later once we add
+		// messaging keys from the DB.
+		if reflect.DeepEqual(messagingKeyEntry.PublicKey[:], publicKey) {
+			messagingKeysMap[*messagingKeyEntry.MessagingPublicKey] = messagingKeyEntry
 			continue
 		}
-		messageKey := MakeMessageKey(publicKey, viewMessageEntry.TstampNanos)
-		if viewMessageKey != messageKey {
-			continue
+		for _, recipient := range messagingKeyEntry.Recipients {
+			if reflect.DeepEqual(recipient.RecipientPublicKey[:], publicKey) {
+				// If user is a recipient of a group chat, we need to add a modified messaging entry.
+				recipientKeyEntry := &MessagingKeyEntry {
+					PublicKey: recipient.RecipientPublicKey,
+					MessagingPublicKey: messagingKeyEntry.MessagingPublicKey,
+					MessagingKeyName: recipient.RecipientMessagingKeyName,
+					EncryptedKey: recipient.EncryptedKey,
+					isDeleted: messagingKeyEntry.isDeleted,
+					isRecipient: true,
+				}
+				messagingKeysMap[*messagingKeyEntry.MessagingPublicKey] = recipientKeyEntry
+				break
+			}
+		}
+	}
+
+	// Also get all messaging keys from the DB.
+	dbMessagingKeys, err := DBGetAllUserMessagingKeys(bav.Handle, publicKey)
+	if err != nil {
+		return nil, errors.Wrapf(err, "GetUserMessagingKeys: problem getting " +
+			"messaging keys from the DB")
+	}
+	// Now go through the messaging keys in the DB and add keys we haven't seen before.
+	for _, messagingKeyEntry := range dbMessagingKeys {
+		key := *messagingKeyEntry.MessagingPublicKey
+		// Check if we have seen the messaging key before.
+		if _, exists := messagingKeysMap[key]; !exists {
+			messagingKeysMap[key] = messagingKeyEntry
+		} else {
+			// If the messaging key was deleted in the UtxoView, we can now delete it from our map.
+			if messagingKeysMap[key].isDeleted {
+				delete(messagingKeysMap, key)
+			}
```

Comment:
Do we want to delete this here? Seems unnecessary and could result in a bug where we've deleted the key from the map before flushing to the db.

Reply:
We're deleting from a local map that only makes reference to UtxoView's map, so it seems like it should be fine to delete here.